src/node_api.cc

```diff
@@ -275,9 +277,32 @@ class ThreadSafeFunction : public node::AsyncResource {
     return napi_ok;
   }
 
-  void DispatchOne() {
+  inline void* Context() {
+    return context;
+  }
+
+ protected:
+  void Dispatch() {
+    bool has_more = true;
+
+    // Limit maximum synchronous iteration count to prevent event loop
+    // starvation. See `src/node_messaging.cc` for an inspiration.
+    unsigned int iterations_left = kMaxIterationCount;
+    while (has_more && --iterations_left != 0) {
+      dispatch_state = kDispatchRunning;
+      has_more = DispatchOne();
+
+      // Send() was called while we were executing the JS function
+      if (dispatch_state.exchange(kDispatchIdle) != kDispatchRunning) {
```

Comment:
Fwiw, you could merge this exchange call with the store call immediately coming up in the next iteration into a compare + exchange call – probably doesn’t make a huge difference in most cases because the calls really follow each other immediately, but it’s something that is going to stand out to people who spend too much time working with atomics :)

Reply:
Oh, that's nice!

Unfortunately, I don't fully get how to make it work specifically here, because I have to know if this loop should continue running before the store above happens. Could you elaborate?

Reply:
Err, yeah, so, there’s already an issue here that I’m only noticing because I’m trying to write up how the cmpxchg variant would look like: When `iterations_left` is about to be reached, `Send()` can still be called before this right `exchange()` call here, assume that it doesn’t have to call `uv_async_send()` because it sees `kDispatchRunning`, but the work doesn’t actually get executed because of the `iterations_left` condition above.

Something like this should work, I think?

```c++

  void Dispatch() {
    // Limit maximum synchronous iteration count to prevent event loop
    // starvation. See `src/node_messaging.cc` for an inspiration.
    unsigned int iterations_left = kMaxIterationCount;
    dispatch_state = kDispatchRunning;
    while (--iterations_left != 0) {
      bool has_more = DispatchOne();
      if (!has_more) {
        // Check if Send() was called while we were executing the JS function
        unsigned char running = kDispatchRunning;
        if (dispatch_state.compare_exchange_strong(running, kDispatchIdle)) {
          break;  // Value set to kDispatchIdle successfully, i.e. no Send().
        }
      }
    }
    if (dispatch_state.exchange(kDispatchIdle) & kDispatchPending) {
      Send();
    }
  }
```

Reply:
Oh, this is a good point. I think we can avoid `uv_async_send()` here by just starting idle though. Does it sound right?

Reply:
Err, yeah, we can. Tbh, I don’t fully understand why there is an `uv_idle_t` here in addition to the `uv_async_t`.

Reply:
I'm thinking doing a simpler change, actually:

```diff
diff --git a/src/node_api.cc b/src/node_api.cc
index 9f3fbdf4fb..374963b46d 100644
--- a/src/node_api.cc
+++ b/src/node_api.cc
@@ -297,6 +297,10 @@ class ThreadSafeFunction : public node::AsyncResource {
         has_more = true;
       }
     }
+
+    if (has_more) {
+      CHECK_EQ(0, uv_idle_start(&idle, IdleCb));
+    }
   }
 
   bool DispatchOne() {
```

Reply:
> Err, yeah, we can. Tbh, I don’t fully understand why there is an `uv_idle_t` here in addition to the `uv_async_t`.

Oh, I didn't understand it at first too. Whoever wrote this was pretty smart. The point is that `uv_async_send()` kicks off the processing, but because multiple calls to it might result in a single `uv_async_cb()` call - `uv_idle_t` keeps ticking until it all queued callbacks are processed. Now that I think about it, it should actually be active even without any change to the code.

Reply:
Right, but you could also just use the `uv_async_t` to re-schedule another callback.

Reply:
So you basically suggest that we get rid of `uv_idle_t` in favor of keeping only `uv_async_t`?

Reply:
@addaleax fixed. PTAL.